### PR TITLE
Cci photogram industrial 2

### PR DESCRIPTION
**Pull request by Cindellai**

### CODE REVIEW

File: app/models/follow_request.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#
+#  index_follow_requests_on_recipient_id  (recipient_id)
+#  index_follow_requests_on_sender_id     (sender_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (recipient_id => users.id)
+#  fk_rails_...  (sender_id => users.id)
+#
+class FollowRequest < ApplicationRecord
```
Comment (kcalva) [inline app/models/follow_request.rb:22]: add enum column type (status) to this model

File: app/models/like.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#  index_likes_on_photo_id  (photo_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (fan_id => users.id)
+#  fk_rails_...  (photo_id => photos.id)
+#
+class Like < ApplicationRecord
+  belongs_to :fan, class_name: "User", counter_cache: true
+  belongs_to :photo, counter_cache: true
```
Comment (kcalva) [inline app/models/like.rb:23]: add validation for users so they won't have more than one like to the photo.

File: app/models/photo.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+# Indexes
+#
+#  index_photos_on_owner_id  (owner_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (owner_id => users.id)
+#
+class Photo < ApplicationRecord
+  belongs_to :owner, class_name: "User", counter_cache: true
```
Comment (kcalva) [inline app/models/photo.rb:26]: if in the Comment and Like model they have a belong to photo, then Photo model should have a hasMany

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  has_many :sent_follow_requests, foreign_key: :sender_id, class_name: "FollowRequest"
+  has_many :accepted_sent_follow_requests, -> { where(status: "accepted") }, foreign_key: :sender_id, class_name: "FollowRequest"
+  has_many :received_follow_requests, foreign_key: :recipient_id, class_name: "FollowRequest"
+  has_many :accepted_received_follow_requests, -> { where(status: "accepted") }, foreign_key: :recipient_id, class_name: "FollowRequest"
+  has_many :likes, foreign_key: :fan_id
```
Comment (kcalva) [inline app/models/user.rb:35]: add dependent::destroy to has many associations need to make sure the Child object is destroyed before the Parent

**Review by kcalva — CHANGES_REQUESTED** [review]
Found some associations missing and other key factors that need to be added

File: app/models/comment.rb
Comment (SaraDawner2000) [inline app/models/comment.rb:1]: Validatations look good

File: app/models/comment.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+#  index_comments_on_photo_id  (photo_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (author_id => users.id)
+#  fk_rails_...  (photo_id => photos.id)
+#
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: "User", counter_cache: true
+  belongs_to :photo
```
Comment (SaraDawner2000) [inline app/models/comment.rb:23]: needs to have counter_cache: true (there is a comments_count column in the Photos table)

**Comment by Cindellai** [comment]
i need to remake it

File: app/models/like.rb
Comment (SaraDawner2000) [inline app/models/like.rb:1]: Thanks, I missed validating that a user hasn't liked the photo already!

File: app/models/photo.rb (hunk 3 above)
Comment (SaraDawner2000) [reply app/models/photo.rb:26]: Missed the has_many likes myself!

File: app/models/user.rb (hunk 4 above)
Comment (SaraDawner2000) [reply app/models/user.rb:35]: Was this in part 2?

File: app/views/comments/_comment.html.erb
Comment (SaraDawner2000) [inline app/views/comments/_comment.html.erb:1]: I see you got started on the frontend?

File: app/views/comments/_form.html.erb
Comment (SaraDawner2000) [inline app/views/comments/_form.html.erb:1]: same as above

File: app/views/comments/edit.html.erb
Comment (SaraDawner2000) [inline app/views/comments/edit.html.erb:1]: same as above

File: app/views/comments/index.html.erb
Comment (SaraDawner2000) [inline app/views/comments/index.html.erb:1]: same as above

File: app/views/comments/new.html.erb
Comment (SaraDawner2000) [inline app/views/comments/new.html.erb:1]: same as above

File: config/routes.rb
Comment (SaraDawner2000) [inline config/routes.rb:1]: good job on catching that root, I would've forgotten to fix it))))

File: lib/tasks/dev.rake [hunk 6]
```diff
@@ -1,3 +1,70 @@
...
+    )
+  end
+  p "#{User.count} users have been created."
+
+
+  users = User.all
+
+  users.each do |first_user|
+    users.each do |second_user|
+      next if first_user == second_user
```
Comment (SaraDawner2000) [inline lib/tasks/dev.rake:28]: oooh, nice catch!

File: lib/tasks/dev.rake [hunk 7]
```diff
@@ -1,3 +1,70 @@
...
+  users.each do |user|
+    rand(15).times do
+      photo = user.own_photos.create(
+        caption: Faker::Quote.jack_handey,
+        image: "https://robohash.org/#{rand(9999)}"
+      )
+
+      user.followers.each do |follower|
+        if rand < 0.5 && !photo.fans.include?(follower)
+          photo.fans << follower
```
Comment (SaraDawner2000) [inline lib/tasks/dev.rake:53]: clever way to add likes!!!!

**Review by SaraDawner2000 — CHANGES_REQUESTED** [review]
Looks great! Minor thing: i think in the comment model we should have counter_cache set to true on the belongs_to :photo association because we have a comment_count column in the Photos table

**Pull request closed without merging**